### PR TITLE
[Feature](datatype) Add IPv4/v6 data type for doris

### LINKED ISSUE #21370: [Feature](datatype) Add  IPv4/v6 data type for doris
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Now We do not support  IPv4/v6 data type for doris witch is common  data type in database
So it's the time to support it ! We have some basic funnctions for data type to check the data type is supported completly

- [ ] support CURD 
- [ ] support StreamLoad
- [ ] support Serde functions 
- [ ] support (implicit/explicit) cast functions 
- [ ] specific functions for itself data type (u can reference clickhouse or mysql)
        eg: IPv4ToIPv6(xxx) 

- [ ] make a Performance Test result with clickhouse or ElasticSearch in load this data type

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by sjyango**
## Proposed changes

Issue Number: close #21370

## Describe your changes
Now We do not support IPv4/v6 data type for doris witch is common data type in database
So it's the time to support it ! We have some basic funnctions for data type to check the data type is supported completly

- [x] support CURD

 - [x] support StreamLoad

 - [x] support Serde functions

 - [x] support (implicit/explicit) cast functions. The following cast functions have been implemented：
1. uint8/int8 -> ipv4
2. uint16/int16 -> ipv4
3. uint32/int32 -> ipv4
4. uint64/int64 -> ipv4
5. string -> ipv4
6. string -> ipv6

- [x] specific functions for itself data type (u can reference clickhouse or mysql). The following functions have been implemented：
1. ipv4numtostring(n)
2. ipv4stringtonum(s)
3. ipv4stringtonum_or_null(s)
4. ipv4stringtonum_or_default(s)
5. inet_ntoa(n)
6. inet_aton(s)
7. isipv4string(s)
8. ipv6numtostring(n)
9. ipv6stringtonum(s)
10. ipv6stringtonum_or_null(s)
11. ipv6stringtonum_or_default(s)
12. inet6_ntoa(n)
13. inet6_aton(s)
14. isipv6string(s)

 - [x] make a Performance Test result with clickhouse or ElasticSearch in load this data type
test case:

A CSV file containing three columns(id int, ip_v4 ipv4, ip_v6 ipv6), with a file size of 400MB.

**Doris: 27.861s**
DDL SQL as follows: 
```SQL
CREATE TABLE IF NOT EXISTS ip_stream_load (
  `id` int,
  `ip_v4` ipv4,
  `ip_v6` ipv6
) ENGINE=OLAP
DISTRIBUTED BY HASH(`id`) BUCKETS 4
PROPERTIES (
"replication_allocation" = "tag.location.default: 1"
);
```
**ClickHouse: 1s**
DDL SQL as follows: 
```SQL
CREATE TABLE IF NOT EXISTS ip_stream_load (
    id Int32,
    ip_v4 IPv4,
    ip_v6 IPv6
) ENGINE = MergeTree()
ORDER BY id;
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 52.13 seconds
 stream load tsv:          501 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          86.8 seconds inserted 10000000 Rows, about 115K ops/s
 storage size: 17168387850 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230708143008_clickbench_pr_174817.html

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 54.55 seconds
 stream load tsv:          505 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          89.4 seconds inserted 10000000 Rows, about 111K ops/s
 storage size: 17166057618 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230709123022_clickbench_pr_175065.html

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 54.91 seconds
 stream load tsv:          497 seconds loaded 74807831229 Bytes, about 143 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          90.2 seconds inserted 10000000 Rows, about 110K ops/s
 storage size: 17169602007 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230710132341_clickbench_pr_175699.html

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run feut

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 51.23 seconds
 stream load tsv:          510 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         18 seconds loaded 2358488459 Bytes, about 124 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.8 seconds inserted 10000000 Rows, about 347K ops/s
 storage size: 17170094862 Bytes

**Comment by sjyango** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 50.12 seconds
 stream load tsv:          504 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         18 seconds loaded 2358488459 Bytes, about 124 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162114219 Bytes

**Comment by sjyango** [comment]
run p0

File: be/src/vec/core/field.h [hunk 1]
```diff
@@ -936,10 +936,6 @@ struct Field::EnumToType<Field::Types::UInt64> {
     using Type = UInt64;
 };
 template <>
-struct Field::EnumToType<Field::Types::UInt128> {
```
Comment (amorynan) [inline be/src/vec/core/field.h:939]: why delete this?

File: be/src/vec/data_types/serde/data_type_ipv4_serde.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+#include <type_traits>
+
+#include "gutil/casts.h"
+#include "vec/columns/column_const.h"
+
+namespace doris {
+namespace vectorized {
+
+void DataTypeIPv4SerDe::write_column_to_arrow(const IColumn& column, const NullMap* null_map,
```
Comment (amorynan) [inline be/src/vec/data_types/serde/data_type_ipv4_serde.cpp:30]: here no need to impl this function, becase it has same behavor with DataTypeNumberSerde<UInt32>

File: be/src/vec/data_types/serde/data_type_ipv6_serde.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+#include <type_traits>
+
+#include "gutil/casts.h"
+#include "vec/columns/column_const.h"
+
+namespace doris {
+namespace vectorized {
+
+void DataTypeIPv6SerDe::write_column_to_arrow(const IColumn& column, const NullMap* null_map,
```
Comment (amorynan) [inline be/src/vec/data_types/serde/data_type_ipv6_serde.cpp:30]: same with IPV4

File: be/src/vec/core/field.h (hunk 1 above)
Comment (sjyango) [reply be/src/vec/core/field.h:939]: This is my mistake, it really shouldn't have been deleted

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run buildall

**Comment by amorynan** [comment]
rebase master let regress pass

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run buildall

**Pull request closed without merging**